### PR TITLE
Enhancement/Azure Gov Cloud Support

### PR DESCRIPTION
**Pull request by nesdxr**
Added an optional --azure-env parameter, options are
AZURE_PUBLIC_CLOUD (default)
AZURE_US_GOV_CLOUD
AZURE_CHINA_CLOUD
AZURE_GERMAN_CLOUD

Tested to work on both AZURE_PUBLIC_CLOUD and AZURE_US_GOV_CLOUD. Should also work for others since this forces passing of the predefined Azure cloud environment.

### CODE REVIEW

**Comment by nesdxr** [comment]
This is my first time to participate. I have no idea what the failed items mean. Would appreciate if someone points me to resources. Thanks.

**Comment by j4v** [comment]
Brilliant @nesdxr, I'll review and merge ASAP.

File: ScoutSuite/providers/azure/facade/graphrbac.py [hunk 1]
```diff
@@ -8,8 +8,8 @@
 class GraphRBACFacade:
     def __init__(self, graphrbac_credentials, credentials, tenant_id, subscription_id):
         self._subscription_id = subscription_id
-        self._client = GraphRbacManagementClient(graphrbac_credentials, tenant_id=tenant_id)
-        self._authorization_client = AuthorizationManagementClient(credentials, subscription_id=subscription_id)
+        self._client = GraphRbacManagementClient(graphrbac_credentials, tenant_id=tenant_id, base_url=graphrbac_credentials.cloud_environment.endpoints.resource_manager)
```
Comment (j4v) [inline ScoutSuite/providers/azure/facade/graphrbac.py]: This is required specifically for other environments? I'm asking as it isn't required in the default cloud environment...

File: ScoutSuite/providers/azure/authentication_strategy.py [hunk 2]
```diff
@@ -40,6 +40,13 @@ def authenticate(self,
...
             logging.getLogger('urllib3').setLevel(logging.ERROR)
 
+            cloud_dic = {"AZURE_PUBLIC_CLOUD":AZURE_PUBLIC_CLOUD,
+                         "AZURE_CHINA_CLOUD":AZURE_CHINA_CLOUD,
+                         "AZURE_US_GOV_CLOUD":AZURE_US_GOV_CLOUD,
+                         "AZURE_GERMAN_CLOUD":AZURE_GERMAN_CLOUD}
+
+            cloud_env = cloud_dic.get(kargs.get('azure_env'), AZURE_PUBLIC_CLOUD)
+
             if cli:
```
Comment (j4v) [inline ScoutSuite/providers/azure/authentication_strategy.py]: No changes are needed for this to work with CLI authentication?

**Comment by j4v** [comment]
@nesdxr looks mostly good, see a few comments (I've also made a few changes, see branch https://github.com/nccgroup/ScoutSuite/tree/nesdxr-bugfix/337-AzureGOV).

File: ScoutSuite/providers/azure/facade/graphrbac.py (hunk 1 above)
Comment (nesdxr) [reply ScoutSuite/providers/azure/facade/graphrbac.py]: Passing the base_url ensures that the proper resource_manager endpoint is used, one of the breakpoints when using Azure gov.

File: ScoutSuite/providers/azure/authentication_strategy.py (hunk 2 above)
Comment (nesdxr) [reply ScoutSuite/providers/azure/authentication_strategy.py]: We have a tight environment that won't let us use Azure CLI. We connect via service principal. The snag in our case is in the call to the ServicePrincipalCredentials.

Comment (j4v) [reply ScoutSuite/providers/azure/authentication_strategy.py]: Looks like it's working for all facades but graphrbac, I'm getting a number of errors:
```
2019-10-18 09:46:54 wrkbx scout[17579] ERROR graphrbac.py L22: Failed to retrieve users: Operation returned an invalid status code 'Not Found'
2019-10-18 09:46:54 wrkbx scout[17579] ERROR graphrbac.py L29: Failed to retrieve groups: Operation returned an invalid status code 'Not Found'
2019-10-18 09:46:54 wrkbx scout[17579] ERROR graphrbac.py L46: Failed to retrieve service principals: Operation returned an invalid status code 'Not Found'
2019-10-18 09:46:54 wrkbx scout[17579] ERROR graphrbac.py L53: Failed to retrieve applications: Operation returned an invalid status code 'Not Found'
```
These disappear if I remove the `base_url` parameter.

Comment (nesdxr) [reply ScoutSuite/providers/azure/authentication_strategy.py]: Thanks for checking. I was able to reproduce. Made the following changes but was not quite able to get pass through it due to "Insufficient privileges" return. I have given most read access under Azure Active Directory and Microsoft Graph but I still get the same message.

The changes I did is:
        self._client = GraphRbacManagementClient(graphrbac_credentials, tenant_id=tenant_id, base_url=graphrbac_credentials.cloud_environment.endpoints.active_directory_graph_resource_id)

so basically using active_directory_graph_resource_id instead of  resource_manager

**Comment by j4v** [comment]
@nesdxr @jagallardo19 could you confirm that the https://github.com/nccgroup/ScoutSuite/tree/nesdxr-bugfix/337-AzureGOV branch properly implements this?

I've tested on a normal subscription with `--user-account` and it's working fine. `--cli` is returning a `'AdalAuthentication' object has no attribute 'cloud_environment'` error, not sure if this is also the case when running against Gov Cloud (?).

**Comment by j4v** [comment]
@nesdxr do you plan on addressing the above, or should we close the PR?

**Comment by j4v** [comment]
Closing due to lack of feedback.
